New Rule L052: Semi colon alignment

jpy-git opened the pull request:
<!--Firstly, thanks for adding this feature! Secondly, please check the key steps against the checklist below to make your contribution easy to merge.-->

<!--Please give the Pull Request a meaningful title (including the dialect this PR is for if it is dialect specific), as this will automatically be added to the release notes, and then the Change Log.-->

### Brief summary of the change made
<!--If there is an open issue for this, then please include `fixes #XXXX` or `closes #XXXX` replacing `XXXX` with the issue number and it will automatically close the issue when the pull request is merged. Alternatively if not fully closed you can say `makes progress on #XXXX` to create a link on that issue without closing it.-->
This PR introduces a new rule L052. It serves two purposes, one to snap semi-colons to the end of the preceding statement, and the other to add the final trailing semi-colon if it is missing:

![Screen-Recording-2021-11-14-at-15 59 30](https://user-images.githubusercontent.com/80432516/141688620-9e9fc29d-15f0-4a30-a663-06a2fc2f95e4.gif)

I have added a configuration option called `allow_final_semi_colon` (which defaults to False) so users can allow missing final semi-colons if they desire (I'm happy to change the default to True on this if you would prefer, but my preference is to add the final semi-colons by default).

This change impacts quite a few unit tests, solely because we weren't previously including final semi-colon, so I have gone through and added the semi-colon/excluded the rule where I think appropriate.

Fixes #1748 

### Are there any other side effects of this change that we should be aware of?
No

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

jpy-git commented:
@barrywhart took a couple attempts to track down all the places this impacted in the dbt tests, but this should be good for review now 😄

WittierDinosaur commented:
I don't think this is an anti-pattern. It's fairly standard in large queries
```
        -- Newlines between the end of the statement
        -- and the semi-colon are forbidden.
        SELECT
            a
        FROM foo
        ;
```

tunetheweb commented:
Personally I like this rule!

What about an optional config to also allow them to be on a newline (with no extra newlines in between) to keep the @WittierDinosaur 's of the world happy too? In which case should this be enforced or only allowed? Which should `fix` use if it's missing and this optional param is set?

jpy-git commented:
@WittierDinosaur @tunetheweb sure thing! I don't think it should be too hard to add that option.

The way it will work is user has an option of either semi-colon on same line OR new line, but not both. Consistentcy should be encouraged.

The default will be to use same line.

alanmcruickshank commented:
So - potential spanner in the works for this rule. Trailing semi colons are explicitly a problem for dbt users: 
![image](https://user-images.githubusercontent.com/4670904/141692191-5777bc56-1c5b-4bbc-8cac-2e7e0f433d3e.png)

Reason being that often queries are "wrapped" by other sql and so the trailing semicolon becomes an issue.

Potential solutions:
- This rule is disabled by default for users when using the dbt templater (although I'm not sure exactly what the mechanism should be to do that).
- That semi-colons _between_ statements are linted by default, but that the default is that _trailing_ semicolons aren't added.

jpy-git commented:
@alanmcruickshank ah was not aware of that, agree that your approach to disable the trailing semi-colon by default is pragmatic though, will tweak the default config 👍

alanmcruickshank commented:
> @alanmcruickshank ah was not aware of that, agree that your approach to disable the trailing semi-colon by default is pragmatic though, will tweak the default config 👍

If we're taking that option - would you mind removing the semi-colons from the tests in the dbt tests. I'm happy for them to stay on the core tests - but I think that sets the right expectation for performance with dbt. Bonus points if you make a note in the default config to that effect.

jpy-git commented:
Will do! 😄

jpy-git commented:
Addressed the "allow_final_semi_colon" being set to true by default point. Will move on to looking at the option for semi-colons on a new line now

jpy-git commented:
@WittierDinosaur @tunetheweb @alanmcruickshank I've added the options to configure the behaviour of rule L052 as we described above. I've also added extra unit tests to for the config options.

Demo:
![semi-colon-config-options](https://user-images.githubusercontent.com/80432516/141700281-55b15c24-d01c-4e49-b4d9-0e0914c3eff8.gif)

jpy-git commented:
@barrywhart Ready for review 😄

barrywhart commented:
Tagging other commenters for review. I'll take a look as well.

barrywhart commented on src/sqlfluff/core/default_config.cfg:
```diff
@@ -43,6 +43,8 @@ comma_style = trailing
 allow_scalar = True
 single_table_references = consistent
 unquoted_identifiers_policy = all
+semi_colon_new_line = False
```
Isn't "semicolon" a single (compound) word? I'm thinking we don't want that underscore there.
```suggestion
semicolon_newline = False
```

jpy-git replied:
I think you can also spell it semi-colon, hence the underscore but happy to do one word 👍 
Should I maybe move that config to a rule specific config as well?

barrywhart replied:
I would move it, yes. It's unlikely another rule would use something so specific.

jpy-git replied:
Flowed these changes through to code and unit tests

barrywhart commented on src/sqlfluff/core/default_config.cfg:
```diff
@@ -43,6 +43,8 @@ comma_style = trailing
 single_table_references = consistent
 unquoted_identifiers_policy = all
+semi_colon_new_line = False
+require_final_semi_colon = False
```
```suggestion
require_final_semicolon = False
```

jpy-git replied:
Flowed these changes through to code and unit tests

tunetheweb reviewed (approved):
LGTM

jpy-git commented:
@tunetheweb @barrywhart think we can merge this, L053, and L054 down once the final checks are done, I've addressed all the suggestions and test cases look good
